lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -4298,9 +4298,21 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
     //     print("Caught bar error")
     //   }
     // }
-    if (fromExistential) {
+    //
+    // Note: we relax the restriction if the type we're casting to is a
+    // non-final class because it's possible that we might have a subclass
+    // that conforms to the protocol.
+    if (fromExistential && !toExistential) {
       if (auto NTD = toType->getAnyNominal()) {
-        if (!isa<ProtocolDecl>(NTD)) {
+        bool shouldContinue = true;
+
+        // If we have a non-final class, then don't check
+        if (toType->is<ClassType>() && !NTD->isFinal()) {
```

Comment:
Shouldn't it just be if (!toType->is<ClassType>() || NTD->isFinal())` so we don't need a boolean flag here?

Reply:
Yeah, you're right!